### PR TITLE
MCOPY for lfvm

### PR DESCRIPTION
**Pull request by LuisPH3**
MCOPY for lfvm
- instruction 
- interpreter test (same semantics as c++)
- No new memory infrastructure, copy is done with a read+write with a stash slice in between. 

Part of #467

### CODE REVIEW

**Comment by LuisPH3** [comment]
Note, CI fails because unrelated issue: https://github.com/Fantom-foundation/Tosca/issues/507
The code here can be reviewed.

File: go/vm/lfvm/instructions.go [hunk 1]
```diff
@@ -206,6 +206,47 @@ func opMstore8(c *context) {
...
+
+	destOffset, destOverflow := destAddr.Uint64WithOverflow()
+	srcOffset, srcOverflow := srcAddr.Uint64WithOverflow()
+	if destOverflow || srcOverflow || !sizeU256.IsUint64() {
+		c.status = ERROR
+		return
+	}
+
+	size := sizeU256.Uint64()
+	price := vm.Gas(3 * ((size + 31) / 32))
```
Comment (HerbertJordan) [inline go/vm/lfvm/instructions.go:234]: `size + 31` may overflow, right?

File: go/vm/lfvm/instructions_test.go [hunk 2]
```diff
@@ -285,3 +286,123 @@ func TestCreateChecksBalance(t *testing.T) {
...
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+
+			ctrl := gomock.NewController(t)
+			ctxt := context{
+				status: RUNNING,
+				params: vm.Parameters{
+					Recipient: vm.Address{1},
+				},
```
Comment (HerbertJordan) [inline go/vm/lfvm/instructions_test.go:382]: This can be removed.
```suggestion
```

File: go/vm/lfvm/instructions_test.go [hunk 3]
```diff
@@ -285,3 +286,123 @@ func TestCreateChecksBalance(t *testing.T) {
...
+				stack:  NewStack(),
+				memory: NewMemory(),
+			}
+			runContext := vm.NewMockRunContext(ctrl)
+			ctxt.context = runContext
```
Comment (HerbertJordan) [inline go/vm/lfvm/instructions_test.go:387]: This is also not relevant for the test.
```suggestion
```

File: go/vm/lfvm/instructions_test.go [hunk 4]
```diff
@@ -285,3 +286,123 @@ func TestCreateChecksBalance(t *testing.T) {
...
+			if ctxt.status != test.expectedStatus {
+				t.Errorf("expected status %v, got %v", test.expectedStatus, ctxt.status)
+			}
+			if ctxt.memory.Len() != uint64(len(test.memoryAfter)) {
+				t.Errorf("expected memory size %d, got %d", uint64(len(test.memoryAfter)), ctxt.memory.Len())
+			}
+			if !bytes.Equal(ctxt.memory.Data(), test.memoryAfter) {
+				t.Errorf("expected memory %v, got %v", test.memoryAfter, ctxt.memory.Data())
+			}
+		})
```
Comment (HerbertJordan) [inline go/vm/lfvm/instructions_test.go:406]: Do you want to add a check that the full gas was consumed?

**Review by HerbertJordan — APPROVED** [review]
Looks good, just some minor things that should be checked before merging.

File: go/vm/lfvm/instructions.go (hunk 1 above)
Comment (LuisPH3) [reply go/vm/lfvm/instructions.go:234]: Yes, I have been thinking a lot about this, nevertheless this is the same code copied all over the place. 

I'll fix it here #512

File: go/vm/lfvm/instructions_test.go [hunk 5]
```diff
@@ -285,3 +286,128 @@ func TestCreateChecksBalance(t *testing.T) {
 	}
 
 }
+func TestMCopy(t *testing.T) {
+
+	tests := map[string]struct {
+		gasBefore      vm.Gas
+		gasAfter       vm.Gas
```
Comment (HerbertJordan) [inline go/vm/lfvm/instructions_test.go:293]: nit: if the parameter is always 0, no need to have it.

**Review by HerbertJordan — APPROVED** [review]

File: go/vm/lfvm/instructions_test.go (hunk 2 above)
Comment (HerbertJordan) [reply go/vm/lfvm/instructions_test.go:385]: This code seems to still be there ...

Comment (LuisPH3) [reply go/vm/lfvm/instructions_test.go]: Sorry about that.

**Review by HerbertJordan — APPROVED** [review]

**Review by facuMH — APPROVED** [review]

**Pull request merged**